okhttp/src/main/kotlin/com/connectrpc/okhttp/ConnectOkHttpClient.kt

```diff
@@ -40,12 +45,49 @@ import java.net.SocketTimeoutException
  * The OkHttp implementation of HTTPClientInterface.
  */
 class ConnectOkHttpClient @JvmOverloads constructor(
-    private val unaryClient: OkHttpClient = OkHttpClient(),
-    private val streamClient: OkHttpClient = unaryClient,
+    unaryClient: OkHttpClient = OkHttpClient(),
+    streamClient: OkHttpClient = unaryClient,
 ) : HTTPClientInterface {
+    private val unaryClient = applyNetworkInterceptor(unaryClient)
+    private val streamClient = applyNetworkInterceptor(streamClient)
+
+    private fun applyNetworkInterceptor(client: OkHttpClient): OkHttpClient {
+        return client.newBuilder()
+            .addNetworkInterceptor(object : Interceptor {
+                override fun intercept(chain: Interceptor.Chain): Response {
+                    val resp = chain.proceed(chain.request())
+                    // The Connect protocol spec currently suggests 408 as the HTTP status code for
+                    // cancelled and deadline exceeded errors with unary methods.. However, the
+                    // spec for this status code states that the request may be retried without
+                    // modification (even non-idempotent requests), and okhttp does in fact retry
+                    // such errors. (So do many browsers.) So if we see that error code AND an
+                    // application/json body (which suggests this is such a Connect error, vs. a
+                    // real client timeout error or a 408 generated by a middlebox), we will
+                    // transform the error in a network interceptor to prevent okhttp from
+                    // retrying it.
+                    if (resp.code == 408 && isConnectUnary(chain.request())) {
+                        val contentType = resp.headers["Content-Type"]
+                        if (contentType == "application/json" || contentType == "application/json; charset=utf-8") {
+                            return resp.newBuilder().code(499).build()
```

Comment:
I guess if you want to restore the 408 status code at the end you could have an interceptor (not network interceptor) that restores it if it was set here.

Reply:
Instead, I've added a way to get the original code and now store that in connect's `HTTPResponse`, and also use it when computing a Connect RPC error code from an HTTP status code. WDYT?

Reply:
Looks great!